data/appRateLimiter.go

```diff
@@ -0,0 +1,283 @@
+		start := time.Now()
+		el := r.lrus[b].Back()
+		delete(r.buckets[b], el.Value)
+		r.lrus[b].Remove(el)
+
+		// some stats for benchmarks/tests, remove?
+		atomic.AddUint64(&r.evictions, 1)
+		atomic.AddUint64(&r.evictionTime, uint64(time.Since(start)))
+	}
+	el := r.lrus[b].PushFront(key)
+	entry := &appRateLimiterEntry{lruElement: el}
+	entry.cur.Store(1)
+	entry.interval.Store(curInt)
+	r.buckets[b][key] = entry
+}
+
+// interval calculates the interval numeric representation based on the given time
+func (r *appRateLimiter) interval(now time.Time) int64 {
+	return now.UnixNano() / int64(r.serviceRateWindow)
+}
+
+// fraction calculates the fraction of the interval that is elapsed since the given time
+func (r *appRateLimiter) fraction(now time.Time) float64 {
+	return float64(now.UnixNano()%int64(r.serviceRateWindow)) / float64(r.serviceRateWindow)
+}
+
+// shouldDrop returns true if the given transaction group should be dropped based on the
+// on the rate for the applications in the group: the entire group is dropped if a single application
+// exceeds the rate.
+func (r *appRateLimiter) shouldDrop(txgroup []transactions.SignedTxn, origin []byte) bool {
+	return r.shouldDropInner(txgroup, origin, time.Now())
+}
+
+// shouldDropInner is the same as shouldDrop but accepts the current time as a parameter
+// in order to make it testable
+func (r *appRateLimiter) shouldDropInner(txgroup []transactions.SignedTxn, origin []byte, now time.Time) bool {
+	buckets, keys := txgroupToKeys(txgroup, origin, r.seed, r.salt, numBuckets)
+	if len(keys) == 0 {
+		return false
+	}
+	return r.shouldDropKeys(buckets, keys, now)
+}
+
+func (r *appRateLimiter) shouldDropKeys(buckets []int, keys []keyType, now time.Time) bool {
+	curInt := r.interval(now)
+
+	for i := range keys {
+		key := keys[i]
+		b := buckets[i]
+		r.mus[b].Lock()
+		entry, has := r.buckets[b][key]
+		if !has {
+			// add a new entry and prune the oldest if needed
+			r.newEntryLocked(b, key, curInt)
+			r.mus[b].Unlock()
+		} else {
+			r.mus[b].Unlock()
+			// optimistic lock-free rate check
+			// expecting that buckets are large enough to avoid contention and frequent evictions
+
+			interval := entry.interval.Load()
+			// copy cur and prev values to check limits
+			// apply limits only after admission
+			cur := entry.cur.Load()
+			prev := entry.prev.Load()
+			thisInterval := true
+			var newPrev uint64 = 0
+			if interval != curInt {
+				thisInterval = false
+				if interval == curInt-1 {
+					// there are continuous intervals, use the previous value
+					newPrev = cur
+				}
+				prev = newPrev
+				cur = 1
+			} else {
+				cur++
+			}
+
+			curFraction := r.fraction(now)
+			rate := uint64(float64(prev)*(1-curFraction)) + cur
+
+			if rate > r.serviceRatePerWindow {
+				return true
+			}
+
+			// admitted, update the entry
+			if thisInterval {
+				entry.cur.Add(1)
+			} else {
+				entry.prev.Store(newPrev)
+				entry.cur.Store(1)
+				entry.interval.Store(curInt)
+			}
```

Comment:
This rolls back the rate violating key but it still doesn't roll back the earlier admitted-now-rejected keys in the range. 
1. maybe prune only once at the end of a successful shouldDropKeys instead of on each newEntryLocked and avoid depruning in rollback and the double-check performed below. 
2. You'll have to keep track of map inserts performed in newEntryLocked and the initial value of each entry (taken before the optimistic unlock) to properly do the rollback.

Reply:
rewrote this by reverting this two locks implementation back to single lock. Incrementing after admission. Updating LRU on attempts.